### ISSUE TITLE
Field xxx has wrong value: Invalid value.

### ISSUE THREAD

**Issue by jackandjackadswithsm** [sangria-graphql/sangria#423]
**Mutation using Object Literal fails with Field xxx has wrong value.**

Could you please let us know if this error is because of unsupported feature, a bug, or something wrong..

We are using a custom scalar for mutation operation and running into issue..

Here are the details of the  scalar data type, schema definition, and mutation (that works using stringified JSON but not using JS Object Literal)

**Custom Scalar Definition**

implicit val configurationDataType = ScalarType[JsValue]("configurationData",
    description = Some("Either Raw JSON or Object literal value"),
    coerceOutput = (value, _) ⇒ value.toString(),
    coerceUserInput = {
      case v: String ⇒ Right(Json.parse(v))
      case v: Boolean ⇒ Right(JsBoolean(v))
      case v: Int ⇒ Right(JsNumber(v))
      case v: Long ⇒ Right(JsNumber(v))
      case v: Float ⇒ Right(JsNumber(BigDecimal(v.toDouble)))
      case v: Double ⇒ Right(JsNumber(v))
      case v: BigInt ⇒ Right(JsNumber(BigDecimal(v)))
      case v: BigDecimal ⇒ Right(JsNumber(v))
      case v: JsValue ⇒ Right(v)
    },
    coerceInput = {
      case ast.StringValue(jsonStr, _, _, _, _) ⇒ {
        Right(Json.parse(jsonStr))
      }
      case ast.ObjectValue(fields,_,_) ⇒ {
          val configurartionPayLoad =  fields.map(data => {
          
          // Some code to convert input data to JSON String using NASHORN Java 8 approach
          // Convert the update input argument data to JSON String
          val result = nashorn.eval(s"var x = $jsString;  JSON.stringify(x)", scriptContext)
          
          val parsedJson = Json.parse(result.toString)
          Json.obj("layers" -> parsedJson)
        })
        Right(configurartionPayLoad.head)
      }
      case _ ⇒ {
        Left(JsonCoercionViolation)
      }

    })

// **Schema Definition**
case class UpdateEntityUnsetInput(configuration: Boolean)

case class UpdateEntitySetInput(configuration: String)

case class UpdateEntityInput(entityId: String,
                             set: Option[UpdateEntitySetInput],
                             unset: Option[UpdateEntityUnsetInput])

@Singleton
class MyMutationSchemaDefinition {

  import play.api.libs.json._
  import sangria.macros.derive._
  import sangria.marshalling.playJson._

  implicit lazy val updateEntityUnSetInputFormat = Json.format[UpdateEntityUnsetInput]
  implicit lazy val updateEntitySetInputFormat = Json.format[UpdateEntitySetInput]
  implicit lazy val updateEntityFormat = Json.format[UpdateEntityInput]

  implicit lazy val UpdateEntitySetInputType = InputObjectType[UpdateEntitySetInput]("UpdateEntitySetInput", List(
    InputField("configuration", controllers.schema.utils.SchemaHelpers.configurationDataType)))

  implicit lazy val UpdateEntityUnsetInputType = deriveInputObjectType[UpdateEntityUnsetInput]()

  implicit lazy val UpdateentityInputType = InputObjectType[UpdateentityInput]("UpdateentityInput", List(
    InputField("entityId", IDType),
    InputField("set", OptionInputType(UpdateEntitySetInputType)),
    InputField("unset", OptionInputType(UpdateEntityUnsetInputType))))

  val UpdateEntityInputArg = Argument("input", UpdateEntityInputType)

}

**Mutation Using JS Object Literal Notation**

mutation {
  updateEntity(input: {entityId: “idValue", 
    set: {configuration: {layers: [{type: "LayerConfiguration", 
      key: {key1: "9155922381708984503", key2: “value2"}, key3: true},
    {type: "LayerConfiguration", 
      key: {key1: “8768312111424122525", key2: "value2"}, key3: true}
    ]}}}) {
    entity {
      id
      configuration {
        data
      }
    }
  }
}

**Response / Output:** Mutation Fails with the following error

{
  "data": null,
  "errors": [
    {
      "message": "Field 'input.set.configuration' has wrong value: Invalid value. (line 5, column 3):\n  updateEntity(input: {entityId: \”idValue\", \n  ^\n (line 6, column 33):\n    set: {configuration: {layers: [{type: \"LayerConfiguration\", \n   ^",
      "path": [
        "updateEntity"
      ],
      "locations": [
        {
          "line": 5,
          "column": 3
        },
        {
          "line": 6,
          "column": 33
        }
      ]
    }
  ]
}

**Mutation Using Escaped JSON String Notation** 

mutation {
  updateEntity(input: {entityId: “idValue", 
    set: {configuration: "{\"layers\":[{\"type\":\"LayerConfiguration\",\"key\":{\"key1\":\"9155922381708984503\",\"key2\":\"value2\"},\"key3\":true}]}"}}) {
    entity {
      id
      configuration {
        data
      }
    }
  }
}

**Response / Output:** Mutation is successful without errors

**Comment by jackandjackadswithsm** [comment]
Using the same code, mutation using Json stringfy works but fails using object literal..

**Comment by marioosh** [comment]
You're trying to read it as JSON, so it have to be JSON. Specification says that every key in JSON is a string so it have to be quoted. 
JS Object literals is a Javascript feature, so it seems like not a bug...

**Comment by jackandjackadswithsm** [comment]
@bengraygh - Here is the formatted output. I have posted the same in Gitter, no response.

**Custom Scalar Definition**

```
implicit val configurationDataType = ScalarType[JsValue]("configurationData",
    description = Some("Either Raw JSON or Object literal value"),
    coerceOutput = (value, _) ⇒ value.toString(),
    coerceUserInput = {
      case v: String ⇒ Right(Json.parse(v))
      case v: Boolean ⇒ Right(JsBoolean(v))
      case v: Int ⇒ Right(JsNumber(v))
      case v: Long ⇒ Right(JsNumber(v))
      case v: Float ⇒ Right(JsNumber(BigDecimal(v.toDouble)))
      case v: Double ⇒ Right(JsNumber(v))
      case v: BigInt ⇒ Right(JsNumber(BigDecimal(v)))
      case v: BigDecimal ⇒ Right(JsNumber(v))
      case v: JsValue ⇒ Right(v)
    },
    coerceInput = {
      case ast.StringValue(jsonStr, _, _, _, _) ⇒ {
        Right(Json.parse(jsonStr))
      }
      case ast.ObjectValue(fields,_,_) ⇒ {
          val configurartionPayLoad =  fields.map(data => {
          
          // Some code to convert input data to JSON String using NASHORN Java 8 approach
          // Convert the update input argument data to JSON String
          val result = nashorn.eval(s"var x = $jsString;  JSON.stringify(x)", scriptContext)
          
          val parsedJson = Json.parse(result.toString)
          Json.obj("layers" -> parsedJson)
        })
        Right(configurartionPayLoad.head)
      }
      case _ ⇒ {
        Left(JsonCoercionViolation)
      }

    })
```
**Schema Definition**
```
case class UpdateEntityUnsetInput(configuration: Boolean)

case class UpdateEntitySetInput(configuration: String)

case class UpdateEntityInput(entityId: String,
                             set: Option[UpdateEntitySetInput],
                             unset: Option[UpdateEntityUnsetInput])

@Singleton
class MyMutationSchemaDefinition {

  import play.api.libs.json._
  import sangria.macros.derive._
  import sangria.marshalling.playJson._

  implicit lazy val updateEntityUnSetInputFormat = Json.format[UpdateEntityUnsetInput]
  implicit lazy val updateEntitySetInputFormat = Json.format[UpdateEntitySetInput]
  implicit lazy val updateEntityFormat = Json.format[UpdateEntityInput]

  implicit lazy val UpdateEntitySetInputType = InputObjectType[UpdateEntitySetInput]("UpdateEntitySetInput", List(
    InputField("configuration", controllers.schema.utils.SchemaHelpers.configurationDataType)))

  implicit lazy val UpdateEntityUnsetInputType = deriveInputObjectType[UpdateEntityUnsetInput]()

  implicit lazy val UpdateentityInputType = InputObjectType[UpdateentityInput]("UpdateentityInput", List(
    InputField("entityId", IDType),
    InputField("set", OptionInputType(UpdateEntitySetInputType)),
    InputField("unset", OptionInputType(UpdateEntityUnsetInputType))))

  val UpdateEntityInputArg = Argument("input", UpdateEntityInputType)

}

```
**Mutation Using JS Object Literal Notation**
```
mutation {
  updateEntity(input: {entityId: “idValue", 
    set: {configuration: {layers: [{type: "LayerConfiguration", 
      source: {source_id: "9155922381708984503", sourceType: “LayerReference"}, visible: true},
    {type: "LayerConfiguration", 
      source: {source_id: “8768312111424122525", sourceType: "LayerReference"}, visible: true}
    ]}}}) {
    entity {
      id
      configuration {
        data
      }
    }
  }
}
```
**Response / Output**: Mutation Fails with the following error
```
{
  "data": null,
  "errors": [
    {
      "message": "Field 'input.set.configuration' has wrong value: Invalid value. (line 5, column 3):\n  updateEntity(input: {entityId: \”idValue\", \n  ^\n (line 6, column 33):\n    set: {configuration: {layers: [{type: \"LayerConfiguration\", \n   ^",
      "path": [
        "updateEntity"
      ],
      "locations": [
        {
          "line": 5,
          "column": 3
        },
        {
          "line": 6,
          "column": 33
        }
      ]
    }
  ]
}
```
**Mutation Using Escaped JSON String Notation** 
```
mutation {
  updateEntity(input: {entityId: “idValue", 
    set: {configuration: "{\"layers\":[{\"type\":\"LayerConfiguration\",\"source\":{\"source_id\":\"9155922381708984503\",\"sourceType\":\"LayerReference\"},\"visible\":true}]}"}}) {
    entity {
      id
      configuration {
        data
      }
    }
  }
}
```
**Response / Output:** Mutation is successful without errors.


@marioosh - I am aware of the difference between JSON and JS Object literal.
Basically I would like to read JS Object literal data.
 Is this supported in Sangria or not ? If so, how should we read ?

**Comment by jackandjackadswithsm** [comment]
I would like to iterate that client will be sending the data in JS Object literal part and the service need to read this.

**Comment by bengraygh** [comment]
Hi, Jack. GraphQL services are typically JSON. Why don't you send JSON to your sangria backed GraphQL service? Presumably, your client is written in javascript, so JSON.stringify(obj) would probably help. But this is very low level, and normally handled by your GraphQL client like Apollo. What is your client?

**Comment by paulpdaniels** [comment]
One side note @jackandjackadswithsm , I'm not sure if this is a problem with the copy and paste you did, but it appears that somehow you are using two different quotation types: `“idValue"` Notice the first quote is different from the second one (The first is ascii: `226 128 156` while the second is `034`). I have a feeling that might also have an effect on the the parser being able to recover the values. Doesn't appear to happen all the time, but it does happen again here: `source_id: “8768312111424122525"`. You may want to fix that as well to make sure it isn't an encoding issue.

**Comment by jackandjackadswithsm** [comment]
@paulpdaniels - Thanks for pointing out..This is copy and paste problem. Please ignore those quotes. 

@bengraygh - As you mentioned correctly, the client is written in Javascript and they strictly follow GraphqlSpec.  The client is sending the data in this format.

Since the spec supports this, the question is 
a) Can Sangria support this? If not, could you please take it as a feature request.
b) As you see in Custom Scala definition, I am able to convert to JSON using JSON.Stringfy with the javascriptEngine. 
````
case ast.ObjectValue(fields,_,_) ⇒ {
          val configurartionPayLoad =  fields.map(data => {
          
          // Some code to convert input data to JSON String using NASHORN Java 8 approach
          // Convert the update input argument data to JSON String
          val result = nashorn.eval(s"var x = $jsString;  JSON.stringify(x)", scriptContext)
          
          val parsedJson = Json.parse(result.toString)
          Json.obj("layers" -> parsedJson)
        })
````
When I debug, the error occurs after this. I tried to debug Sangria code to see why I am getting "Mutation using Object Literal fails with Field xxx has wrong value." error, but couldn't trace.

**Comment by bengraygh** [comment]
Hi Jack.

a). I don't think it does. Again, I don't really understand why you want this, rather than writing client code to create JSON. Why don't you give sangria the format it expects? It does not appear that anyone else is looking for this capability, and I don't see what would be achieved by it.
b). I wish you luck finding the error, but think you should change the client instead.

**Comment by jackandjackadswithsm** [comment]
@bengraygh and @OlegIlyenko - Could you please provide your inputs. If this (JS Object Literal) is not supported , could you please take this as a feature request as GraphQL Spec defines this feature.

Here is the GraphQL Spec I was referring about.
https://facebook.github.io/graphql/draft/#sec-Input-Object-Values

**Comment by OlegIlyenko** [comment]
@jackandjackadswithsm Sangria is fully compliant to the most recent version of the spec. When it comes to JSON scalar type, it is not part of the spec. At the moment several libraries allow to define such scalar types, but it is important to note that it is not standardized and different libraries represent it different ways. One can also [define JSON scalar type with Sangria](https://gist.github.com/OlegIlyenko/5b96f4b54f656aac226d3c4bc33fd2a6), though it also requires adjustment to the `ResultMarshaller` and `InputUnmarshaller`. Also, I think it is important to note that GraphQL input object literal (the one you referenced from the spec) is not the same as JSON object literal, even though they might look similar at a first glance (for example one can use variable and enum literal as a part of definition, there is no such concept in JSON). 

In general, I would suggest to use custom JSON scalar type with caution. While the output behaviour is relatively straightforward to define, the input side might be quite ambiguous and error-prone (especially when it comes to the default values and introspection). I think the most safe way to represent it is to use a string literal. For example:

```graphql
mutation {
  updateEntity(
    input: {
      entityId: “idValue", 
      set: {configuration: "{\"layers": [{\"type\": \"LayerConfiguration\", ...}]}"}
    }
  ) {
    entity {
      id
      configuration {
        data
      }
    }
  }
}
```
